### PR TITLE
InplaceStringBuilder: better performance

### PR DESCRIPTION
**Pull request by gfoidl**
# Description

* covered current behavior with tests + added some test to cover the public api
* instead of throwing exceptions directly, they are thrown via _ThrowHelper_ or private methods, thus making the code for the method smaller
* decreased the size of the struct, by removing two fields (`_capacity` and `_writing`)
* `_value` gets initialized in the ctor, potentially reinitialized in setting the _Capacity_
* made _Append_-methods to inline

A variant with `char[]` instead of `string` as `_value` was tried, but the results didn't satisfy.

# Benchmarks

## Append String, StringSegment and Char

[Benchmark-Code](https://github.com/gfoidl/Benchmarks/blob/a954945e44ec2b936076de51106b5968e9fecfd9/aspnet/Common/InplaceStringBuilder/InplaceStringBuilder/Benchmarks/AppendRealisticBenchmarks.cs)

``` ini

BenchmarkDotNet=v0.10.12, OS=Windows 10 Redstone 3 [1709, Fall Creators Update] (10.0.16299.125)
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical cores and 4 physical cores
Frequency=2742191 Hz, Resolution=364.6719 ns, Timer=TSC
.NET Core SDK=2.1.4
  [Host]     : .NET Core 2.0.5 (Framework 4.6.26020.03), 64bit RyuJIT
  DefaultJob : .NET Core 2.0.5 (Framework 4.6.26020.03), 64bit RyuJIT


```
|                                 Method |      Mean |     Error |    StdDev | Scaled | ScaledSD |
|--------------------------------------- |----------:|----------:|----------:|-------:|---------:|
|                                Default |  74.63 ns | 1.5745 ns | 2.5426 ns |   1.00 |     0.00 |
| Default_ThrowHelper_AggressiveInlining |  42.83 ns | 0.5936 ns | 0.5262 ns |   0.57 |     0.02 |
|                    Default_ThrowHelper |  54.69 ns | 0.3238 ns | 0.2704 ns |   0.73 |     0.02 |
|                              CharArray | 178.26 ns | 1.7530 ns | 1.6397 ns |   2.39 |     0.08 |

## Append String

[Benchmark-Code](https://github.com/gfoidl/Benchmarks/blob/a954945e44ec2b936076de51106b5968e9fecfd9/aspnet/Common/InplaceStringBuilder/InplaceStringBuilder/Benchmarks/AppendStringBenchmarks.cs)

``` ini

BenchmarkDotNet=v0.10.12, OS=Windows 10 Redstone 3 [1709, Fall Creators Update] (10.0.16299.125)
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical cores and 4 physical cores
Frequency=2742191 Hz, Resolution=364.6719 ns, Timer=TSC
.NET Core SDK=2.1.4
  [Host]     : .NET Core 2.0.5 (Framework 4.6.26020.03), 64bit RyuJIT
  DefaultJob : .NET Core 2.0.5 (Framework 4.6.26020.03), 64bit RyuJIT


```
|                                 Method |       Mean |     Error |    StdDev | Scaled | ScaledSD |
|--------------------------------------- |-----------:|----------:|----------:|-------:|---------:|
|                                Default |   453.2 ns |  8.977 ns | 14.749 ns |   1.00 |     0.00 |
| Default_ThrowHelper_AggressiveInlining |   278.1 ns |  4.503 ns |  4.212 ns |   0.61 |     0.02 |
|                    Default_ThrowHelper |   340.9 ns |  4.625 ns |  4.100 ns |   0.75 |     0.03 |
|                              CharArray | 1,451.3 ns | 19.951 ns | 16.660 ns |   3.21 |     0.11 |

## Append Char

[Benchmark-Code](https://github.com/gfoidl/Benchmarks/blob/a954945e44ec2b936076de51106b5968e9fecfd9/aspnet/Common/InplaceStringBuilder/InplaceStringBuilder/Benchmarks/AppendCharBenchmarks.cs)

``` ini

BenchmarkDotNet=v0.10.12, OS=Windows 10 Redstone 3 [1709, Fall Creators Update] (10.0.16299.125)
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical cores and 4 physical cores
Frequency=2742191 Hz, Resolution=364.6719 ns, Timer=TSC
.NET Core SDK=2.1.4
  [Host]     : .NET Core 2.0.5 (Framework 4.6.26020.03), 64bit RyuJIT
  DefaultJob : .NET Core 2.0.5 (Framework 4.6.26020.03), 64bit RyuJIT


```
|              Method |     Mean |     Error |    StdDev | Scaled | ScaledSD |
|-------------------- |---------:|----------:|----------:|-------:|---------:|
|             Default | 553.2 ns | 10.889 ns | 18.193 ns |   1.00 |     0.00 |
| Default_ThrowHelper | 242.9 ns |  3.939 ns |  3.684 ns |   0.44 |     0.02 |
|           CharArray | 213.0 ns |  4.288 ns |  8.158 ns |   0.39 |     0.02 |

# Discussion

## AggressiveInlining

Perf with _aggressive inling_ is a lot better than without. So I'll kept it, although the caller might get bigger in code size.

## Initialization of _value

`_value` gets initialized in the ctor, when a capacity is given. If a new capacity is set via the property, so a new string is initialized, throwing away a potential previous initialized string (_value). This simplifies the checks while appending.

The current version initialized the `_value` on its first use, so there is no throw-away even if the capacity is set by the property.

I think the "new bahavior" is it worth, though setting the capacity via ctor and later via property may be a less common approach.
Or in other words: when the capacity is set via ctor xor set via property, there is no throw-away either.

### CODE REVIEW

**Comment by rynowak** [comment]
@pakrym - are you most familiar with this type? want to take a look at this **SPICY** pr?

File: src/Microsoft.Extensions.Primitives/InplaceStringBuilder.cs [hunk 1]
```diff
@@ -10,98 +10,123 @@ namespace Microsoft.Extensions.Primitives
...
-                    throw new ArgumentOutOfRangeException(nameof(value));
+                    ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.value);
                 }
-                if (_writing)
+
+                // _offset > 0 indicates writing state
+                if (_offset > 0)
                 {
-                    throw new InvalidOperationException("Cannot change capacity after write started.");
+                    ThrowHelper.ThrowInvalidOperationException(ExceptionResource.Capacity_CannotChangeAfterWriteStarted);
```
Comment (pakrym) [inline src/Microsoft.Extensions.Primitives/InplaceStringBuilder.cs]: Why did you move this exception message, but not the other to the resources?

**Comment by pakrym** [comment]
I like this, my only concern is the following pattern:

```
var sb = new InplaceStringBuilder(5);
if (...)
sb.Capacity += 7;
if (...)
sb.Capacity += 3;
```

I looked through asp.net core code and it seems we don't use it but if anyone did things would get much slower.

How expensive is the `if ()` check in append?

File: src/Microsoft.Extensions.Primitives/InplaceStringBuilder.cs [hunk 2]
```diff
@@ -10,98 +10,123 @@ namespace Microsoft.Extensions.Primitives
...
-            if (_offset + length > _capacity)
+
+            fixed (char* destination = _value)
             {
-                throw new InvalidOperationException($"Not enough capacity to write '{length}' characters, only '{_capacity - _offset}' left.");
+                destination[_offset++] = c;
             }
         }
 
+        private void ThrowNotEnoughCapacity(int length)
```
Comment (pakrym) [inline src/Microsoft.Extensions.Primitives/InplaceStringBuilder.cs]: Group throw methods

File: src/Microsoft.Extensions.Primitives/InplaceStringBuilder.cs (hunk 1 above)
Comment (gfoidl) [reply src/Microsoft.Extensions.Primitives/InplaceStringBuilder.cs]: For me the `ThrowHelper` is a reusable piece of code.

`ThrowValidationError` is special to the `InplaceStringBuilder` and accesses `Capacity`, thats way this is kept local.

`ThrowNotEnoughCapacity` and `ThrowCapacityNotUsed` are also specialized for the `InplaceStringBuilder` with access to local data. Plus the string is built at runtime and not so easy to represent as enum -- though still possible.

Or did you mean different exception messages?

Comment (pakrym) [reply src/Microsoft.Extensions.Primitives/InplaceStringBuilder.cs]: I mean exception message moved to resources.

Comment (gfoidl) [reply src/Microsoft.Extensions.Primitives/InplaceStringBuilder.cs]: Sorry, I don't know which ones you mean. Either they have `ExceptionArgument` or the string is built at runtime. Can you tell me the line-no?

**Comment by gfoidl** [comment]
> How expensive is the `if ()` check in append?

This depends on the usage pattern. Depending on that the JIT will or will not inline some methods and the property setter (for Capacity), resulting in +5% to -20% compared to the code of this PR.

### Append String

|              Method |     Mean |    Error |   StdDev | Scaled | ScaledSD |
|-------------------- |---------:|---------:|---------:|-------:|---------:|
|             [Default](https://github.com/gfoidl/Benchmarks/blob/39a3ea8621a6d0885a7ca6c8981a10e5583b34b8/aspnet/Common/InplaceStringBuilder/InplaceStringBuilder/InplaceStringBuilders/InplaceStringBuilder0.cs) | 425.6 ns | 3.895 ns | 3.041 ns |   1.00 |     0.00 |
|             [This_PR](https://github.com/gfoidl/Benchmarks/blob/39a3ea8621a6d0885a7ca6c8981a10e5583b34b8/aspnet/Common/InplaceStringBuilder/InplaceStringBuilder/InplaceStringBuilders/InplaceStringBuilder00.cs) | 282.4 ns | 5.741 ns | 5.370 ns |   0.66 |     0.01 |
| [This_PR_wo_pre_init](https://github.com/gfoidl/Benchmarks/blob/39a3ea8621a6d0885a7ca6c8981a10e5583b34b8/aspnet/Common/InplaceStringBuilder/InplaceStringBuilder/InplaceStringBuilders/InplaceStringBuilder02.cs) | 291.5 ns | 5.920 ns | 4.944 ns |   0.69 |     0.01 |
|            [New_Idea](https://github.com/gfoidl/Benchmarks/blob/39a3ea8621a6d0885a7ca6c8981a10e5583b34b8/aspnet/Common/InplaceStringBuilder/InplaceStringBuilder/InplaceStringBuilders/InplaceStringBuilder01.cs) | 264.2 ns | 5.662 ns | 6.520 ns |   0.62 |     0.02 |

### Append String, StringSegment and Char

|              Method |     Mean |     Error |    StdDev | Scaled |  Gen 0 | Allocated |
|-------------------- |---------:|----------:|----------:|-------:|-------:|----------:|
|             [Default](https://github.com/gfoidl/Benchmarks/blob/39a3ea8621a6d0885a7ca6c8981a10e5583b34b8/aspnet/Common/InplaceStringBuilder/InplaceStringBuilder/InplaceStringBuilders/InplaceStringBuilder0.cs) | 68.31 ns | 0.7611 ns | 0.6747 ns |   1.00 | 0.0151 |      48 B |
|             [This_PR](https://github.com/gfoidl/Benchmarks/blob/39a3ea8621a6d0885a7ca6c8981a10e5583b34b8/aspnet/Common/InplaceStringBuilder/InplaceStringBuilder/InplaceStringBuilders/InplaceStringBuilder00.cs) | 44.31 ns | 0.3620 ns | 0.3023 ns |   0.65 | 0.0152 |      48 B |
| [This_PR_wo_pre_init](https://github.com/gfoidl/Benchmarks/blob/39a3ea8621a6d0885a7ca6c8981a10e5583b34b8/aspnet/Common/InplaceStringBuilder/InplaceStringBuilder/InplaceStringBuilders/InplaceStringBuilder02.cs) | 34.11 ns | 0.4492 ns | 0.4202 ns |   0.50 | 0.0152 |      48 B |
|            [New_Idea](https://github.com/gfoidl/Benchmarks/blob/39a3ea8621a6d0885a7ca6c8981a10e5583b34b8/aspnet/Common/InplaceStringBuilder/InplaceStringBuilder/InplaceStringBuilders/InplaceStringBuilder01.cs) | 34.92 ns | 0.2254 ns | 0.1998 ns |   0.51 | 0.0152 |      48 B |

`New_Idea` is this PR + tracking of the capacity as field, thus allowing the JIT to inline the access to `Capacity`. 

The differences in perf results from the JIT's inlining behavior.

My favorite is now `New_Idea`, but this has the same concern as https://github.com/aspnet/Common/pull/313#issuecomment-367395930
As stated in the description of the PR, I don't think this does matter because the throw-away of the pre-init string won't produce a lot garbage -- under the assumption that the capacity isn't set frequently and to big values.

The variant without the pre-initialization is fine, too. Just depending on the usage-pattern it may be a little slower than a pre-init variant.

So what to do?

**Comment by pakrym** [comment]
Let's go with lazy init option, I don't want to throw people that increment `Capacity` off the cliff.

File: src/Microsoft.Extensions.Primitives/InplaceStringBuilder.cs (hunk 1 above)
Comment (pakrym) [reply src/Microsoft.Extensions.Primitives/InplaceStringBuilder.cs]: Oh, I see, it's because it's built at runtime. Let's go with the ThrowHelper pattern corefx uses and put exceptions there and in resources (https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/ThrowHelper.cs)

Comment (gfoidl) [reply src/Microsoft.Extensions.Primitives/InplaceStringBuilder.cs]: A similar pattern as the linked one is already in use in this `ThrowHelper`, except support for older runtimes, that don't have the no-inline behavior for method that don't return.

I'll move these exceptions to the ThrowHelper. The move to the resources is also good, so no string-resources are in the code directly.

File: src/Microsoft.Extensions.Primitives/Resources.resx [hunk 3]
```diff
@@ -120,4 +120,13 @@
   <data name="Argument_InvalidOffsetLength" xml:space="preserve">
     <value>Offset and length are out of bounds for the string or length is greater than the number of characters from index to the end of the string.</value>
   </data>
+  <data name="Capacity_CannotChangeAfterWriteStarted" xml:space="preserve">
+    <value>Cannot change capacity after write started.</value>
+  </data>
+  <data name="Capacity_not_used_entirely" xml:space="preserve">
+    <value>Entire reserved capacity was not used. Capacity: '{0}', written '{1}'.</value>
+  </data>
+  <data name="Not_enough_capacity" xml:space="preserve">
```
Comment (pakrym) [inline src/Microsoft.Extensions.Primitives/Resources.resx]: PascalCase

**Comment by pakrym** [comment]
Looks good, thank you!
